File: config/src/main/java/com/typesafe/config/impl/ConfigNodeInclude.java
package com.typesafe.config.impl;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.List;

import com.typesafe.config.parser.ConfigNode;
import com.typesafe.config.parser.ConfigNodeVisitor;

final class ConfigNodeInclude extends AbstractConfigNode implements com.typesafe.config.parser.ConfigNodeInclude {
    final private ArrayList<AbstractConfigNode> children;
    final private ConfigIncludeKind kind;
    final private boolean isRequired;

    ConfigNodeInclude(Collection<AbstractConfigNode> children, ConfigIncludeKind kind, boolean isRequired) {
        this.children = new ArrayList<AbstractConfigNode>(children);
        this.kind = kind;
        this.isRequired = isRequired;
    }

    final public Collection<AbstractConfigNode> children() {
        return children;
    }

    @Override
    protected Collection<Token> tokens() {
        ArrayList<Token> tokens = new ArrayList<Token>();
        for (AbstractConfigNode child : children) {
            tokens.addAll(child.tokens());
        }
        return tokens;
    }

    protected ConfigIncludeKind kind() {
        return kind;
    }

    protected boolean isRequired() {
        return isRequired;
    }

    protected String name() {
        for (AbstractConfigNode n : children) {
            if (n instanceof ConfigNodeSimpleValue) {
                return (String)Tokens.getValue(((ConfigNodeSimpleValue) n).token()).unwrapped();
            }
        }
        return null;
    }

    @Override
    public <T> T accept(ConfigNodeVisitor<T> visitor) {
        return visitor.visitInclude(this);
    }

    @Override
    public List<ConfigNode> getChildren() {
        return Collections.unmodifiableList(children);
    }
}

File: config/src/main/scala/com/typesafe/config/impl/ConfigNodeInclude.scala
package com.typesafe.config.impl

import java.{ util => ju }

final class ConfigNodeInclude(
    final val children: ju.Collection[AbstractConfigNode],
    protected val kind: ConfigIncludeKind,
    protected val isRequired: Boolean) extends AbstractConfigNode {

    override def tokens: ju.Collection[Token] = {
        val tokens = new ju.ArrayList[Token]
        import scala.collection.JavaConversions._
        for (child <- children) { tokens.addAll(child.tokens) }
        tokens
    }

    protected def name: String = {
        import scala.collection.JavaConverters._
        for (n <- children.asScala) {
            if (n.isInstanceOf[ConfigNodeSimpleValue])
                return Tokens
                    .getValue(n.asInstanceOf[ConfigNodeSimpleValue].token)
                    .unwrapped
                    .asInstanceOf[String]
        }
        null
    }
}
